### ISSUE TITLE
Softmaxsamples broken with move to 1.0

### ISSUE THREAD

**Issue by eschnou** [Conchylicultor/DeepQA#67]
Using --softmaxSamples leads to the error below. I haven't investigated yet on root cause, but a lot of APIs have changed in 1.0.

Traceback (most recent call last):
  File "/usr/local/lib/python3.4/dist-packages/tensorflow/python/framework/common_shapes.py", line 670, in _call_cpp_shape_fn_impl
    status)
  File "/usr/lib/python3.4/contextlib.py", line 66, in __exit__
    next(self.gen)
  File "/usr/local/lib/python3.4/dist-packages/tensorflow/python/framework/errors_impl.py", line 469, in raise_exception_on_not_ok_status
    pywrap_tensorflow.TF_GetCode(status))
tensorflow.python.framework.errors_impl.InvalidArgumentError: Shape must be rank 2 but is rank 1 for 'sequence_loss/sequence_loss_by_example/sampled_softmax_loss/LogUniformCandidateSampler' (op: 'LogUniformCandidateSampler') with input shapes: [?].

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "main.py", line 29, in <module>
    chatbot.main()
  File "/home/eschenal/Workspace/personal/DeepQA/chatbot/chatbot.py", line 166, in main
    self.model = Model(self.args, self.textData)
  File "/home/eschenal/Workspace/personal/DeepQA/chatbot/model.py", line 103, in __init__
    self.buildNetwork()
  File "/home/eschenal/Workspace/personal/DeepQA/chatbot/model.py", line 191, in buildNetwork
    softmax_loss_function= sampledSoftmax if outputProjection else None  # If None, use default SoftMax
  File "/usr/local/lib/python3.4/dist-packages/tensorflow/contrib/legacy_seq2seq/python/ops/seq2seq.py", line 1110, in sequence_loss
    softmax_loss_function=softmax_loss_function))
  File "/usr/local/lib/python3.4/dist-packages/tensorflow/contrib/legacy_seq2seq/python/ops/seq2seq.py", line 1067, in sequence_loss_by_example
    crossent = softmax_loss_function(target, logit)
  File "/home/eschenal/Workspace/personal/DeepQA/chatbot/model.py", line 138, in sampledSoftmax
    self.textData.getVocabularySize()),  # The number of classes
  File "/usr/local/lib/python3.4/dist-packages/tensorflow/python/ops/nn_impl.py", line 1191, in sampled_softmax_loss
    name=name)
  File "/usr/local/lib/python3.4/dist-packages/tensorflow/python/ops/nn_impl.py", line 947, in _compute_sampled_logits
    range_max=num_classes)
  File "/usr/local/lib/python3.4/dist-packages/tensorflow/python/ops/candidate_sampling_ops.py", line 134, in log_uniform_candidate_sampler
    seed2=seed2, name=name)
  File "/usr/local/lib/python3.4/dist-packages/tensorflow/python/ops/gen_candidate_sampling_ops.py", line 357, in _log_uniform_candidate_sampler
    name=name)
  File "/usr/local/lib/python3.4/dist-packages/tensorflow/python/framework/op_def_library.py", line 763, in apply_op
    op_def=op_def)
  File "/usr/local/lib/python3.4/dist-packages/tensorflow/python/framework/ops.py", line 2397, in create_op
    set_shapes_for_outputs(ret)
  File "/usr/local/lib/python3.4/dist-packages/tensorflow/python/framework/ops.py", line 1757, in set_shapes_for_outputs
    shapes = shape_func(op)
  File "/usr/local/lib/python3.4/dist-packages/tensorflow/python/framework/ops.py", line 1707, in call_with_requiring
    return call_cpp_shape_fn(op, require_shape_fn=True)
  File "/usr/local/lib/python3.4/dist-packages/tensorflow/python/framework/common_shapes.py", line 610, in call_cpp_shape_fn
    debug_python_shape_fn, require_shape_fn)
  File "/usr/local/lib/python3.4/dist-packages/tensorflow/python/framework/common_shapes.py", line 675, in _call_cpp_shape_fn_impl
    raise ValueError(err.message)
ValueError: Shape must be rank 2 but is rank 1 for 'sequence_loss/sequence_loss_by_example/sampled_softmax_loss/LogUniformCandidateSampler' (op: 'LogUniformCandidateSampler') with input shapes: [?].

**Comment by Conchylicultor** [comment]
It seems that the parameters `input` and `labels` have been inverted.
https://www.tensorflow.org/api_docs/python/tf/nn/sampled_softmax_loss

So it breaks the code [here](https://github.com/Conchylicultor/DeepQA/blob/master/chatbot/model.py#L135)

**Comment by Conchylicultor** [comment]
In addition, I think a problem is that before, it was possible to use 1d tensor `input` as for `sparse_softmax_cross_entropy_with_logits`. It seems it's not possible anymore with the new `sampled_softmax_loss`.

**Comment by alanbekker** [comment]
Can you provide some tips to how fix it?

**Comment by eschnou** [comment]
I had a try at fixing it, but encountered further issues down the path. It just seems that the new tf.nn.sampled_softmax_loss is not compatible with the argument softmax_loss_function of the legacy seq2seq. Unfortunately, there is no legacy_rnn package.

Two options: 
1. Copy locally the code from tf 0.12, but not sure this can solve everything
2. Migrate to the new seq2seq implementation instead of using the legacy one.

I think 2 is the way forward and will try to invest some time doing it in the coming days. 

@Conchylicultor Did you already have a look at the new API ?

**Comment by alanbekker** [comment]
I managed to solve it by myself, I just inverted the input and labels argument inside the sampledSoftmax function as following:

            def sampledSoftmax(inputs,labels):
                tmp=inputs
                inputs=labels
                labels=tmp
                labels = tf.reshape(labels, [-1, 1])  # Add one dimension (nb of true classes, here 1)

* For some reason even though it's working i didn't saw any run time improvements...(even when I used the numSamples to be 10 on a 40k vocabulary) someone has any idea why?

**Comment by Conchylicultor** [comment]
@alanbekker Indeed the inputs, labels parameters were inverted. Thanks for noticing. I pushed a fix so it sampled softmax should now works. Concerning the performances, it's not clear how sampledSoftmax really helps so I don't have an answer sorry.

@eschnou Yes migrating to the new seq2seq API is definitely something which needs to be done. I did not yet deeply looked into that but I should have a little more time in two weeks.
